src/core/net/dnssd_server.cpp

```diff
@@ -145,8 +148,22 @@ void Server::ProcessQuery(const Header &aRequestHeader, Message &aRequestMessage
     Message         *responseMessage = nullptr;
     Header           responseHeader;
     NameCompressInfo compressInfo(kDefaultDomainName);
-    Header::Response response                = Header::kResponseSuccess;
-    bool             resolveByQueryCallbacks = false;
+    Header::Response response           = Header::kResponseSuccess;
+    bool             shouldSendResponse = true;
+
+#if OPENTHREAD_CONFIG_DNS_UPSTREAM_QUERY_ENABLE
+    if (mEnableUpstreamQuery && ShouldForwardToUpstream(aRequestHeader, aRequestMessage))
+    {
+        error = ResolveByUpstream(aRequestMessage, aMessageInfo);
```

Comment:
True.

BTW, `otUdpReceive` does not mention the ownership of message and we probable need one.

Reply:
I think general rule for the APIs is if it does not transfer ownership, it does not mention it in the doc, but if it does then it is mentioned in the doc (hopefully). I agree improving the docs to make this clear is for sure helpful :)

This management of who owns a message and when we should free is generally complex to track and reason about. A long time back I tried in PR to use smart pointer style to help with this. The downside was that it added a lot of code size overhead. At some point, we may revisit it again. It may be ok to pay the code size overhead price over the simplicity and safety it can provide.